services/system/Transact/plugin/src/helpers.rs

```diff
@@ -33,16 +33,15 @@ pub fn get_sender_app() -> String {
 
 pub fn get_action_sender(service: &str, method: &str) -> Result<String, Host::types::Error> {
     if let Some(plugin) = ActionSenderHook::get() {
-        ActionSenderHook::clear();
```

Comment:
I think this interface creates a high risk of applying the sender to actions that it isn't intended for. Is there ever a case where the caller of `add_action_to_transaction` doesn't know the correct sender?

Reply:
This is the tradeoff of using a hook to set the sender so the actions are always added with the correct sender. The set action sender hook used to only apply to the next action added, but if we are using it for staged tx, then it needs to be persistent so that all actions added downstream of the propose latch have their sender set correctly.

The caller of `add_action_to_transaction` typically does not care who the sender is, they let transact manage it

Reply:
In the case of `branding` "all actions added downstream" is wrong. It's only those specific actions that should use `branding` as the sender.

Reply:
Well, there are no actions added downstream because branding doesn't call any other plugins. If it did, then we would want all those added downstream to be part of the staged-tx.

What should actually happen is that when the plugin who sets the latch pops off the callstack, the latch is unset. But without a mechanism to do that, I suppose it should be standard to call unlatch() at the end?
So branding should be:
```
latch()
add_action_to_tx()
unlatch()
```

Thoughts?

Reply:
Making the sender effectively dynamically scoped would address my concerns here. rust has RAII, so the unlatch doesn't need to be explicit.

Reply:
But it would be up to the plugin to build the raii wrapper for the latch. Or are you suggesting something that I'm missing?

Reply:
Oh right, it's importing wit, not a regular library.

Reply:
I think in the future I might be able to define the RAII in the staged-tx plugin itself, using a [resource](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md#item-resource).

But resources are currently not supported by our wasm component host.

Reply:
Added the ability for `staged-tx` to unhook from `transact` by calling `set_propose_latch(None)` and added RAII to `branding` to automatically manage hook/unhook.
Added comment in staged-tx as a reminder to move the RAII into staged-tx when resources are supported.

Reply:
This works, but it relies on the plugins cooperating. I'm concerned about the security implications if they don't.

For example, let alice be the current user who has her own account, and also has signing power for bob.
```
mallory::evil()
tokens::transfer(carol, 100 PSI)
```

I would expect that mallory could derail the whole thing. I would also expect mallory to be able to add new actions that are sent to mallory's service. I would not expect mallory to be able to change the transfer to come from bob.

I don't think a resource completely solves the problem, because a plugin could choose to keep the resource alive longer than it's supposed to.

Reply:
Okay, I just decided to make staged-tx privileged so it can import `get-active-app` from the host. That allows it to restrict `set_propose_latch` to only the active application or its plugin. There is therefore no problem if the latch is not explicitly unset because the transact hook will expire after the transaction context.

Furthermore, for security, I added some explicit (but temporary) checks in the transact plugin hooks to whitelist the apps that are allowed to hook. We can relax this once we have the client-side oauth stuff, but this keeps the implementation safe in the meantime.